keras/models.py

```diff
@@ -22,6 +22,59 @@
 from .layers import containers
 
 
+def count_named_params(weights):
+    ctr = 0
+    for k, v in weights.items():
+        if isinstance(v, dict):
+            ctr += count_named_params(v)
+        elif isinstance(v, list):
+            ctr += len(v)
+        else:
+            raise ValueError('unrecognized weight structure of type'
+                             .format(type(v)))
+    return ctr
+
+
+def mk_hdf5_tree(root, weights):
+    '''Recursive function to save a weight dictionary to
+    an hdf5 file. To be used in conjuction with `save_named_weights`.
+    '''
+    for k, w in weights.items():
+        subroot = root.create_group(k)
+        if isinstance(w, dict):
+            mk_hdf5_tree(subroot, w)
+        else:
+            subroot.attrs['nb_params'] = len(w)
+            for n, param in enumerate(w):
+                param_name = 'param_{}'.format(n)
+                param_dset = subroot.create_dataset(param_name, param.shape,
+                                                    dtype=param.dtype)
+                param_dset[:] = param
+
+
+def read_hdf5_tree(root):
+    '''Inverse of mk_hdf5 tree. Should be called with the 'graph' Group,
+    not the file root
+    '''
+    # using visititems to add to dict would be more cumbersome
+    # and less direct than crawling directly because of how
+    # visititems stringifies names
+
+    assert root.__class__.__name__ == 'Group'
+
+    if all(v.__class__.__name__ == 'Dataset' for v in root.values()):
+        weights = [root['param_{}'.format(p)]
+                   for p in range(root.attrs['nb_params'])]
```

Comment:
You should be able to get this list without relying on `nb_params`. Then you can get rid of all code related to it.